session/logind.c

```diff
@@ -315,22 +318,35 @@ static struct wlr_session *logind_session_start(struct wl_display *disp) {
 	snprintf(session->base.seat, sizeof(session->base.seat), "%s", seat);
 	free(seat);
 
-	const char *fmt = "/org/freedesktop/login1/session/%s";
-	int len = snprintf(NULL, 0, fmt, session->id);
-
-	session->path = malloc(len + 1);
-	if (!session->path) {
+	ret = sd_bus_default_system(&session->bus);
+	if (ret < 0) {
+		wlr_log(L_ERROR, "Failed to open D-Bus connection: %s", strerror(-ret));
 		goto error;
 	}
 
-	sprintf(session->path, fmt, session->id);
+	ret = sd_bus_call_method(session->bus,
+							 "org.freedesktop.login1",
+							 "/org/freedesktop/login1",
+							 "org.freedesktop.login1.Manager",
+							 "GetSession",
+							 &error,
+							 &msg,
+							 "s",
+							 session->id);
```

Comment:
The indent style here is wrong.
Also, can you move this into its own function like SessionActivate and TakeControl are?